server/src/test/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzerTests.java

```diff
@@ -0,0 +1,526 @@
+            return field.name;
+        }
+    }
+
+    static void rewriteIndexWithPerFieldCodec(Directory source, CodecMode mode, Directory dst) throws IOException {
+        try (DirectoryReader reader = DirectoryReader.open(source)) {
+            IndexWriterConfig config = new IndexWriterConfig()
+                .setSoftDeletesField(Lucene.SOFT_DELETES_FIELD)
+                .setUseCompoundFile(randomBoolean())
+                .setCodec(new Lucene87Codec(mode.mode()) {
+                    @Override
+                    public PostingsFormat getPostingsFormatForField(String field) {
+                        return new Lucene84PostingsFormat();
+                    }
+
+                    @Override
+                    public DocValuesFormat getDocValuesFormatForField(String field) {
+                        return mode.dvFormat();
+                    }
+
+                    @Override
+                    public String toString() {
+                        return super.toString();
+                    }
+                })
+                .setOpenMode(IndexWriterConfig.OpenMode.CREATE);
+            try (IndexWriter writer = new IndexWriter(dst, config)) {
+                for (LeafReaderContext leaf : reader.leaves()) {
+                    final SegmentReader segmentReader = Lucene.segmentReader(leaf.reader());
+                    writer.addIndexes(segmentReader);
+                }
+                writer.commit();
+            }
+        }
+    }
+
+    static IndexDiskUsageStats collectPerFieldStats(Directory directory) throws IOException {
+        try (DirectoryReader reader = DirectoryReader.open(directory)) {
+            final IndexDiskUsageStats stats = new IndexDiskUsageStats(IndexDiskUsageAnalyzer.getIndexSize(lastCommit(directory)));
+            for (LeafReaderContext leaf : reader.leaves()) {
+                collectPerFieldStats(Lucene.segmentReader(leaf.reader()), stats);
+            }
+            return stats;
+        }
+    }
+
+    static void collectPerFieldStats(SegmentReader reader, IndexDiskUsageStats stats) throws IOException {
+        final SegmentInfo sis = reader.getSegmentInfo().info;
+        final String[] files;
+        final Directory directory;
+        if (sis.getUseCompoundFile()) {
+            directory = sis.getCodec().compoundFormat().getCompoundReader(reader.directory(), sis, IOContext.READ);
+            files = directory.listAll();
+        } else {
+            directory = reader.directory();
+            files = sis.files().toArray(new String[0]);
+        }
+        final FieldLookup fieldLookup = new FieldLookup(reader.getFieldInfos());
+        try {
+            for (String file : files) {
+                final String ext = IndexFileNames.getExtension(file);
+                if (ext == null) {
+                    continue;
+                }
+                final long bytes = directory.fileLength(file);
+                switch (ext) {
+                    case "dvd":
```

Comment:
can you reuse LuceneFilesExtensions to not have to update this file when we change file extensions?

Reply:
++, see https://github.com/elastic/elasticsearch/pull/74051/commits/3a6f202680d16ad1ec5bb24fe65b66e9e56fbba3